api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto

```diff
@@ -183,44 +207,211 @@ message JwtHeader {
   string value_prefix = 2;
 }
 
-// This is the Envoy HTTP filter config for JWT authentication.
-// [#not-implemented-hide:]
-message JwtAuthentication {
-  // List of JWT rules to valide.
-  repeated JwtRule rules = 1;
+// This message specifies a Jwt requirement
+//
+// For example: following match will match all requests.
+//
+// .. code-block:: yaml
+//
+//  # not required with an empty message
+//
+//  # require A
+//  provider_id: "iss-A"
+//
+//  # require A or B
+//  requires_any:
+//    requirements:
+//      - provider_id: "iss-A"
+//      - provider_id: "iss-B"
+//
+//  # require A and B
+//  requires_all:
+//    requirements:
+//      - provider_id: "iss-A"
+//      - provider_id: "iss-B"
+//
+//  # require A and (B or C)
+//  requires_all:
+//    requirements:
+//      - provider_id: "iss-A"
+//      - requires_any:
+//        requirements:
+//          - provider_id: "iss-B"
+//          - provider_id: "iss-C"
+//
+//  # require A or (B and C)
+//  requires_any:
+//    requirements:
+//      - provider_id: "iss-A"
+//      - requires_all:
+//        requirements:
+//          - provider_id: "iss-B"
+//          - provider_id: "iss-C"
+//
+message JwtRequirement {
+  oneof requires_type {
+    // Specify a required provider id.
+    string provider_id = 1;
```

Comment:
Make this as a message and including `provider_name`, `audiences`. You are likely to have same issuer with different audiences at different route. So moving audience to this level will make the provider config more reusable.

Reply:
As you are aware that newly added provider_name field allows users to create multiple providers with different audiences for the same issuer.  I feel adding audiences in the JwtRequirment is adding another layer of override, it is introducing too much complexity in term of implementation and easy to use.  Let us not do that for now,  we can add it later if it is really necessary.
@llchan what do you think?

Reply:
Yes I am aware of that you can create multiple provider for same issuer. Moving the audience from each provider to here, so you don't have to do that for each route that requires just different audience from same issuer.

Reply:
If we move audiences from Provider to ProviderRequirement,  it means removing the one in Provider,  it will have problem for Istio Authn.   Based on your idea of "empty rules" means "allow_missing_and_fail = true" for authn.    It will not have any rules,  so it will not have ProviderRequirements defined,   you could not specify "audience" to verify.

Reply:
Let me clarify, I meant in whole, sth like this:

```proto

message JwtRule { // better name wanted...
  string provider_name;
  repeated string audiences;
}

message NoRequirement {
}

message JwtRequirement {
  oneof require_type {
    NoRequirement no_require;
    JwtRule rule;
    JwtRequirementOrList requires_any;
    JwtRequirementAndList requires_all;
  }
}
```

Reply:
I understood your idea.  If we MOVE audiences to JwtRule,  for Istio authn usage, it will not define any JwtRequirement, so it will not define any JwtRule.  If audiences are only in JwtRule, they will not be verified.